engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -443,5 +431,54 @@ object EngineJobExecutionActor {
     *
     * More refinement may appear via #1377.
     */
-  private val factorySingletonRouters = scala.collection.mutable.Map[BackendLifecycleActorFactory, ActorRef]()
+  private var factoryFileHashingRouters = Map[BackendLifecycleActorFactory, ActorRef]()
+
+  /**
+    * Returns a RoundRobinPool of actors based on the backend factory.
+    *
+    * @param backendName                  Name of the backend.
+    * @param backendLifecycleActorFactory A backend factory.
+    * @param actorRefFactory              An actor factory.
+    * @return a RoundRobinPool of actors based on backend factory.
+    */
+  private def factoryFileHashingRouter(backendName: String,
+                                       backendLifecycleActorFactory: BackendLifecycleActorFactory,
+                                       actorRefFactory: ActorRefFactory): ActorRef = {
+    val (originalOrUpdated, result) = getOrElseUpdated(
```

Comment:
That does an extra lookup. This is an `var immutable.Map` version of the `val mutable.Map.getOrElseUpdate` without the second pass for lookups.

See http://stackoverflow.com/questions/4385976/idiomatic-get-or-else-update-for-immutable-map#answer-5840119